### DISCUSSION TITLE
SMTP over TLS in Docker using Nginx Reverse Proxy

### DISCUSSION THREAD

**Discussion by MadeByGabe** [pocketbase/pocketbase#3797] [Q&A] [answered]
Hello everyone,

I am encountering an issue related to SMTP over TLS while running Pocketbase in a Docker container alongside other applications. All these applications are served from the same domain using an Nginx reverse proxy (also in Docker).

This runs in a docker container:

```bash
./pocketbase serve --http=0.0.0.0:8090
```

My Nginx reverse proxy configuration looks like this:

```nginx
location /pb {
  rewrite /pb/(.*) /$1  break;
  proxy_pass http://<%-s.localDomain%>:8090;
  ...
}
```

The problem arises when I attempt to send test emails via SMTP, consistently resulting in an "unencrypted connection" error. I suspect this error stems from serving Pocketbase with only the --http flag, causing it to assume that TLS is not available. However, I'm uncertain about how to incorporate the https flag or the domain attribute, as all my attempts have led to Nginx 502 errors. I pretty much spent a day trying out everything but my knowledge of nginx is very limited.  

Currently, Pocketbase is accessible at an address like this https://example.com/pb. But I'm unable to set up this domain using the serve command because then I can't proxy pass to it (or just don't know how).

I already use another mailer solution on the same server with the same SMTP so this problem wouldn't be a huge issue by itself but I opted for Pocketbase primarily for its built-in authentication but I haven't been able to figure out how to access, for instance, an email verification token or a password reset token without actually sending emails so I can't use the other mail solution for these things.

I appreciate any guidance or suggestions you can provide to resolve this issue. Please let me know if you need additional details or clarification on any aspect of the problem.

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure that I understand your setup, nor how to reproduce it. If you have nginx/certbot to take care for the certificate, using just the `--http` flag to specify the local port should work and no extra configurations for the TLS should be needed.

Could you paste the exact error message?

What SMTP provider do you use?

Are you sure that you are using the correct SMTP port and/or configuration?

Can you reproduce the issue with the test smtp server of [mailtrap.io](https://mailtrap.io/email-sandbox/) (_they have free plan_)?

**Reply by MadeByGabe** [reply]
The error comes from [here](https://github.com/pocketbase/pocketbase/blob/985ab1e5b79a328b516decca3d1d7dc8edfcbba0/tools/mailer/smtp.go#L172).  I couldn't follow it completely but it seemed to me that those checks are the direct result of me not using the --https flag or the domain attribute.  
The smtp provider is [Acumbamail](https://acumbamail.com/en/) which allows connection without tls also. With ports like 25252 I can send emails through this provider without tls but pocketbase has a timeout error with these ports. The ports which force tls give the mentioned error message.  
I try out mailtrap in a minute.

**Reply by ganigeorgiev** [reply]
PocketBase doesn't really have an option to send an email without TLS, except if you count StartTLS.

I cannot find documentaiton about Acumbamail SMTP but try with `587` port instead.

**Reply by ganigeorgiev** [reply]
Also why do you use `LOGIN` auth instead of `PLAIN`?
The `LOGIN` auth mechanism is usually required by Microsoft. For everything else try first with `PLAIN`.

**Reply by MadeByGabe** [reply]
With mailtrap two ports work (587 and 2525) which I assume the non tls ports and the other two gets me a timeout and those are the tls ports. This is weird and I don't really understand anything but at least I know now that the current configuration at least with some smtp servers works.

**Reply by MadeByGabe** [reply]
I use the plain, but I tried pretty much every combination of every settings. I think it must be e problem with the cert. Currently there is the app which contains pocketbase and another docker container has the nginx config (I use caprover and that works like this), the actual cert files are shared but maybe the clusterfuck with these settings can cause this problem.  
Thanks for the help, I don't think this is actually a pocketbase problem since there is an smtp server I can send emails with.

**Reply by MadeByGabe** [reply]
I don't think I can solve this problem with acumbamail but at this point I have to use that. Can you please help me with hooks or other solutions to replace the PocketBase mail in case of the email verification and password reset? Is that even possible? I looked at all the hooks but I have no idea how to get nodemailer to send those emails.  
Sorry this is a slightly different question, I should probably make another thread for it.

**Reply by MadeByGabe** [reply]
Marked your message with the other smtp service as the actual answer because as it turns out the fault is completely because of [acumbamail](https://appsumo.com/products/acumbamail/reviews/mixed-bag-good-in-many-ways-but-not-sec-1339242/) (tldr, no tls with acumba at all). Sorry for wasting your time. I guess I'm just fked.

**Reply by ganigeorgiev** [reply]
Yes, if the above linked review is correct and Acumbamail SMTP doesn't support TLS connections, then it will not work out of the box with PocketBase because the [Go std PlainAuth](https://pkg.go.dev/net/smtp#PlainAuth) implementation requires the connection to be using TLS or `localhost`. We can create a custom/hacky PlainAuth for this on the Go side but I don't think it will be the correct action in this case and it would be better the provider to add support for TLS to their SMTP server.

Unfortunately there is no easy way to replace in the JS `pb_hooks` the default mailer and I'm not sure if the nodemailer package can be loaded in the hooks. In general you can intercept the mailer hooks and prevent further execution by simply returning `false`:

```js
onMailerBeforeRecordResetPasswordSend((e) => {
    console.log(e.mailClient)
    console.log(e.message)
    console.log(e.record)

    // do something...

    return false;
})
```
